// @lifi/sdk/src/_cjs/core/EVM/permits/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.daiPermitTypes = exports.eip2612Types = exports.TOKEN_ADDRESSES_WITH_SALT = exports.DAI_LIKE_PERMIT_TYPEHASH = exports.EIP712_DOMAINS_WITHOUT_VERSION = exports.EIP712_DOMAIN_TYPEHASH_WITH_SALT = exports.EIP712_DOMAIN_TYPEHASH = exports.DAI_PERMIT_SELECTOR = exports.EIP_2612_PERMIT_SELECTOR = exports.InstantExpiration = exports.MaxSigDeadline = exports.MaxUnorderedNonce = exports.MaxSignatureTransferAmount = exports.MaxOrderedNonce = exports.MaxAllowanceExpiration = exports.MaxAllowanceTransferAmount = void 0;
const constants_js_1 = require("../../../constants.js");
exports.MaxAllowanceTransferAmount = constants_js_1.MaxUint160;
exports.MaxAllowanceExpiration = constants_js_1.MaxUint48;
exports.MaxOrderedNonce = constants_js_1.MaxUint48;
exports.MaxSignatureTransferAmount = constants_js_1.MaxUint256;
exports.MaxUnorderedNonce = constants_js_1.MaxUint256;
exports.MaxSigDeadline = constants_js_1.MaxUint256;
exports.InstantExpiration = 0n;
exports.EIP_2612_PERMIT_SELECTOR = '0xd505accf';
exports.DAI_PERMIT_SELECTOR = '0x8fcbaf0c';
exports.EIP712_DOMAIN_TYPEHASH = '0x8b73c3c69bb8fe3d512ecc4cf759cc79239f7b179b0ffacaa9a75d522b39400f';
exports.EIP712_DOMAIN_TYPEHASH_WITH_SALT = '0x36c25de3e541d5d970f66e4210d728721220fff5c077cc6cd008b3a0c62adab7';
exports.EIP712_DOMAINS_WITHOUT_VERSION = [
    '0x797cfab58fcb15f590eb8e4252d5c228ff88f94f907e119e80c4393a946e8f35',
    '0x8cad95687ba82c2ce50e74f7b754645e5117c3a5bec8151c0726d5857980a866',
];
exports.DAI_LIKE_PERMIT_TYPEHASH = '0xea2aa0a1be11a07ed86d755c93467f4f82362b452371d1ba94d1715123511acb';
exports.TOKEN_ADDRESSES_WITH_SALT = [
    '0x2791bca1f2de4661ed88a30c99a7a9449aa84174:137',
    '0x8f3cf7ad23cd3cadbd9735aff958023239c6a063:137',
];
exports.eip2612Types = {
    Permit: [
        { name: 'owner', type: 'address' },
        { name: 'spender', type: 'address' },
        { name: 'value', type: 'uint256' },
        { name: 'nonce', type: 'uint256' },
        { name: 'deadline', type: 'uint256' },
    ],
};
exports.daiPermitTypes = {
    Permit: [
        { name: 'holder', type: 'address' },
        { name: 'spender', type: 'address' },
        { name: 'nonce', type: 'uint256' },
        { name: 'expiry', type: 'uint256' },
        { name: 'allowed', type: 'bool' },
    ],
};
