// @lifi/sdk/src/_cjs/core/EVM/uns/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CHAIN_TYPE_FAMILY_MAP = exports.CHAIN_ID_UNS_CHAIN_MAP = exports.CHAIN_TYPE_UNS_CHAIN_MAP = exports.UNSProxyReaderABI = exports.getUNSProxyAddress = exports.UNS_PROXY_READER_ADDRESSES = void 0;
const types_1 = require("@lifi/types");
exports.UNS_PROXY_READER_ADDRESSES = {
    [types_1.ChainId.ETH]: '0x578853aa776Eef10CeE6c4dd2B5862bdcE767A8B',
    [types_1.ChainId.POL]: '0x91EDd8708062bd4233f4Dd0FCE15A7cb4d500091',
};
const getUNSProxyAddress = (chainId) => exports.UNS_PROXY_READER_ADDRESSES[chainId];
exports.getUNSProxyAddress = getUNSProxyAddress;
exports.UNSProxyReaderABI = [
    {
        constant: true,
        inputs: [
            {
                internalType: 'string[]',
                name: 'keys',
                type: 'string[]',
            },
            {
                internalType: 'uint256',
                name: 'tokenId',
                type: 'uint256',
            },
        ],
        name: 'getData',
        outputs: [
            {
                internalType: 'address',
                name: 'resolver',
                type: 'address',
            },
            {
                internalType: 'address',
                name: 'owner',
                type: 'address',
            },
            {
                internalType: 'string[]',
                name: 'values',
                type: 'string[]',
            },
        ],
        payable: false,
        stateMutability: 'view',
        type: 'function',
    },
    {
        inputs: [{ internalType: 'uint256', name: 'tokenId', type: 'uint256' }],
        name: 'exists',
        outputs: [{ internalType: 'bool', name: '', type: 'bool' }],
        stateMutability: 'view',
        type: 'function',
    },
];
exports.CHAIN_TYPE_UNS_CHAIN_MAP = {
    [types_1.ChainType.EVM]: 'ETH',
    [types_1.ChainType.MVM]: 'SUI',
    [types_1.ChainType.SVM]: 'SOL',
    [types_1.ChainType.UTXO]: 'BTC',
    [types_1.ChainType.TVM]: 'TRON',
};
exports.CHAIN_ID_UNS_CHAIN_MAP = {
    [types_1.ChainId.ETH]: 'ETH',
    [types_1.ChainId.BTC]: 'BTC',
    [types_1.ChainId.SUI]: 'SUI',
    [types_1.ChainId.SOL]: 'SOL',
    [types_1.ChainId.BAS]: 'BASE',
    [types_1.ChainId.POL]: 'MATIC',
    [types_1.ChainId.ARB]: 'ARB1',
    [types_1.ChainId.AVA]: 'AVAX',
};
exports.CHAIN_TYPE_FAMILY_MAP = {
    [types_1.ChainType.EVM]: 'EVM',
    [types_1.ChainType.UTXO]: 'BTC',
    [types_1.ChainType.SVM]: 'SOL',
    [types_1.ChainType.MVM]: 'SUI',
    [types_1.ChainType.TVM]: 'TRON',
};
