// @bigmi/core/dist/cjs/transports/ankr/getTransactions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTransactions = void 0;
const url_js_1 = require("../../utils/url.js");
const utils_js_1 = require("./utils.js");
const getTransactions = async (client, { baseUrl }, { address, limit = 50, offset = 0, lastBlock }) => {
    const apiUrl = (0, url_js_1.urlWithParams)(`${baseUrl}/address/${address}`, {
        details: 'txs',
        pageSize: limit,
        to: lastBlock,
    });
    const response = (await client.request({
        url: apiUrl,
        fetchOptions: { method: 'GET' },
    }));
    if (response.error) {
        return { error: response.error };
    }
    const total = response.txs || 0;
    const page = Math.floor(offset / limit) + 1;
    const hasMore = total > offset + response.transactions.length;
    const data = {
        transactions: response.transactions.map(utils_js_1.ankrTransactionTransformer),
        total,
        page,
        itemsPerPage: limit,
        hasMore,
    };
    return { result: data };
};
exports.getTransactions = getTransactions;
