// @bigmi/core/dist/cjs/transports/blockchair/getBalance.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBalance = void 0;
const rpc_js_1 = require("../../errors/rpc.js");
const url_js_1 = require("../../utils/url.js");
const getBalance = async (client, { baseUrl, apiKey }, { address }) => {
    const apiUrl = (0, url_js_1.urlWithParams)(`${baseUrl}/addresses/balances`, {
        key: apiKey,
        addresses: address,
    });
    const response = (await client.request({
        url: apiUrl,
        fetchOptions: { method: 'GET' },
    }));
    if (response.data[address] === undefined ||
        response.context.error ||
        response.context?.code !== 200) {
        return {
            error: {
                code: response.context.code === 429
                    ? rpc_js_1.RpcErrorCode.ACCESS_DENIED
                    : rpc_js_1.RpcErrorCode.MISC_ERROR,
                message: response.context?.error,
            },
        };
    }
    return {
        result: BigInt(response.data[address]),
    };
};
exports.getBalance = getBalance;
