// @bigmi/core/dist/cjs/transports/mempool/getTransactions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getTransactions = void 0;
const url_js_1 = require("../../utils/url.js");
const utils_js_1 = require("./utils.js");
const getTransactions = async (client, { baseUrl }, { address, limit = 50, offset = 0, afterTxId }) => {
    const apiUrlAddress = `${baseUrl}/address/${address}`;
    const balanceResponse = (await client.request({
        url: apiUrlAddress,
        fetchOptions: { method: 'GET' },
    }));
    const totalTxns = balanceResponse.chain_stats.tx_count +
        balanceResponse.mempool_stats.tx_count;
    const apiUrl = (0, url_js_1.urlWithParams)(`${baseUrl}/address/${address}/txs`, {
        after_txid: afterTxId,
    });
    const response = (await client.request({
        url: apiUrl,
        fetchOptions: { method: 'GET' },
    }));
    const transactions = response.map(utils_js_1.mempoolTransactionTransformer);
    const result = {
        transactions,
        total: totalTxns,
        itemsPerPage: limit,
        hasMore: offset + transactions.length < totalTxns,
    };
    return {
        result,
    };
};
exports.getTransactions = getTransactions;
