// @elizaos/openclaw-adapter/dist/src/memory-builder.js
import { randomUUID } from "node:crypto";
export function buildMemory(params) {
    const content = { text: params.text, source: params.source ?? "openclaw-adapter" };
    if (params.action)
        content.actions = [params.action];
    return {
        id: randomUUID(),
        content,
        entityId: params.entityId ?? randomUUID(),
        agentId: params.agentId,
        roomId: params.roomId ?? randomUUID(),
        worldId: params.worldId,
        createdAt: Date.now(),
        metadata: { type: "message", source: "openclaw-adapter" },
    };
}
export function buildMemoryFromToolParams(params, actionName, agentId) {
    const parts = [];
    for (const [key, value] of Object.entries(params)) {
        if (value !== undefined && value !== null && value !== "") {
            parts.push(`${key}: ${String(value)}`);
        }
    }
    return buildMemory({
        text: parts.length > 0 ? parts.join(", ") : actionName,
        agentId,
        action: actionName,
        source: "openclaw-adapter",
    });
}
