// @elizaos/openclaw-adapter/dist/index.js
import { parseAdapterConfig } from "./src/config.js";
import { RuntimeBridge } from "./src/runtime-bridge.js";
import { adaptActionToTool } from "./src/action-to-tool.js";
import { adaptProviderToHook } from "./src/provider-to-hook.js";
import { adaptService } from "./src/service-adapter.js";
import { adaptRoute } from "./src/route-adapter.js";
import { adaptEvaluatorToHook } from "./src/evaluator-to-hook.js";
import { mapElizaEventToOpenClawHook } from "./src/event-mapper.js";
async function loadElizaPlugin(specifier, logger) {
    logger.info(`[eliza-adapter] Loading: ${specifier}`);
    const mod = (await import(specifier));
    if (mod["default"] && isPlugin(mod["default"]))
        return mod["default"];
    if (mod["plugin"] && isPlugin(mod["plugin"]))
        return mod["plugin"];
    for (const v of Object.values(mod)) {
        if (isPlugin(v))
            return v;
    }
    throw new Error(`[eliza-adapter] "${specifier}" has no valid Eliza Plugin export.`);
}
function isPlugin(v) {
    return !!v && typeof v === "object" && typeof v["name"] === "string" && typeof v["description"] === "string";
}
async function registerElizaPlugin(plugin, bridge, api) {
    const r = { pluginName: plugin.name, toolCount: 0, hookCount: 0, serviceCount: 0, routeCount: 0 };
    // Start services immediately — Eliza plugins call getService() during init
    for (const sc of plugin.services ?? []) {
        const adapted = adaptService(sc, bridge);
        await adapted.start({ stateDir: "", logger: bridge.logger, config: {} });
        api.registerService({ id: adapted.id, start() { }, stop: adapted.stop });
        r.serviceCount++;
    }
    if (plugin.init) {
        await plugin.init((api.pluginConfig?.["settings"] ?? {}), bridge);
    }
    for (const action of plugin.actions ?? []) {
        const t = adaptActionToTool(action, bridge);
        api.registerTool(t, { name: t.name });
        bridge.registerAction(action);
        r.toolCount++;
    }
    for (const prov of plugin.providers ?? []) {
        const a = adaptProviderToHook(prov, bridge);
        api.on(a.hookName, a.handler);
        bridge.registerProvider(prov);
        r.hookCount++;
    }
    for (const ev of plugin.evaluators ?? []) {
        const a = adaptEvaluatorToHook(ev, bridge);
        api.on(a.hookName, a.handler);
        bridge.registerEvaluator(ev);
        r.hookCount++;
    }
    for (const route of plugin.routes ?? []) {
        const a = adaptRoute(route, bridge);
        if (a) {
            api.registerHttpRoute(a);
            r.routeCount++;
        }
    }
    for (const [eventName, handlers] of Object.entries(plugin.events ?? {})) {
        if (!handlers)
            continue;
        const hookName = mapElizaEventToOpenClawHook(eventName);
        if (!hookName) {
            api.logger.debug?.(`[eliza-adapter] Unmapped event: ${eventName}`);
            continue;
        }
        for (const handler of handlers) {
            api.on(hookName, async () => { await handler({}); });
            r.hookCount++;
        }
    }
    return r;
}
export default {
    id: "eliza-adapter",
    name: "Eliza Plugin Adapter",
    description: "Wraps Eliza plugins as OpenClaw extensions",
    async register(api) {
        const config = parseAdapterConfig(api.pluginConfig);
        api.logger.info(`[eliza-adapter] Loading ${config.plugins.length} plugin(s): ${config.plugins.join(", ")}`);
        const bridge = new RuntimeBridge({ config, openclawLogger: api.logger });
        await bridge.initialize();
        let tools = 0, hooks = 0, services = 0, routes = 0;
        for (const spec of config.plugins) {
            const plugin = await loadElizaPlugin(spec, api.logger);
            const r = await registerElizaPlugin(plugin, bridge, api);
            tools += r.toolCount;
            hooks += r.hookCount;
            services += r.serviceCount;
            routes += r.routeCount;
            api.logger.info(`[eliza-adapter] "${plugin.name}": ${r.toolCount}T ${r.hookCount}H ${r.serviceCount}S ${r.routeCount}R`);
        }
        api.logger.info(`[eliza-adapter] Ready: ${tools}T ${hooks}H ${services}S ${routes}R`);
        api.registerService({ id: "eliza-adapter-lifecycle", start() { }, stop: () => bridge.stop() });
    },
};
